Resolve emoji autocomplete not being temporally consistent

commonlawfeature opened the pull request:
<!-- Please read https://github.com/matrix-org/matrix-js-sdk/blob/develop/CONTRIBUTING.md before submitting your pull request -->

<!-- Include a Sign-Off as described in https://github.com/matrix-org/matrix-js-sdk/blob/develop/CONTRIBUTING.md#sign-off -->

<!-- To specify text for the changelog entry (otherwise the PR title will be used):
Notes: 

Changes in this project generate changelog entries in element-web by default.
To suppress this:

element-web notes: none

...or to specify different notes:
element-web notes: <notes>
-->

Notes: Fix a bug with emoji autocomplete sorting where adding the final "&#58;" would cause the emoji with the typed shortcode to no longer be at the top of the autocomplete list.

Fixes https://github.com/vector-im/element-web/issues/19302

As this is a bugfix, the `why` is probably best addressed by referencing the related issue (https://github.com/vector-im/element-web/issues/19302).

I have included tests in a separate precursor commit so that it is easy to validate that the issue is fixed by the following commit.

Type: defect


<!-- CHANGELOG_PREVIEW_START -->
---
Here's what your changelog entry will look like:

## 🐛 Bug Fixes
 * Fix a bug with emoji autocomplete sorting where adding the final "&#58;" would cause the emoji with the typed shortcode to no longer be at the top of the autocomplete list. ([\#8086](https://github.com/matrix-org/matrix-react-sdk/pull/8086)). Fixes vector-im/element-web#19302. Contributed by @commonlawfeature.<!-- CHANGELOG_PREVIEW_END -->

<!-- Replace -->
Preview: https://pr8086--matrix-react-sdk.netlify.app
⚠️ Do you trust the author of this PR? Maybe this build will steal your keys or give you malware. Exercise caution. Use test accounts.
<!-- Replace -->

commonlawfeature commented on test/autocomplete/EmojiProvider-test.js:
```diff
@@ -0,0 +1,60 @@
+        expect(firstCompletionWithoutColon).toEqual(firstCompletionWithColon);
+    });
+
+    it.each(TOO_SHORT_EMOJI_SHORTNAME)("Returns correct results after final colon %s", async ({ emojiShortcode, expectedEmoji }) => {
```
This line is not correct for the style, but I am not certain how to fix it. I tried matching the style in `matrix-js-sdk/spec/unit/crypto/CrossSigningInfo.spec.js` but that does not seem to be correct for the style in this repo. So, could someone point me toward the correct style to be matching?

SimonBrandner replied:
The line is just too long, perhaps something like this could work:

```suggestion
    it.each(TOO_SHORT_EMOJI_SHORTNAME)("Returns correct results after final colon %s", async (
        { emojiShortcode, expectedEmoji }
    ) => {
```

commonlawfeature replied:
Thanks! Got that resolved.

commonlawfeature commented on src/autocomplete/EmojiProvider.tsx:
```diff
@@ -62,6 +67,19 @@ function score(query, space) {
+        returned = returned.slice(0, -1);
+    }
+    return returned;
+}
```
Per the examples included in the bug reports, some of the emoji include ':' in the middle of the shortcode. Doing that would fundamentally change the searched-for shortcode, which was part of the initial bug. I have tried to be very precise in removing only exactly the characters that have been added.

MadLittleMods replied:
Which emoji have `:` in the middle?

`:man:_beard:` and `:woman:_beard:` was a different bug https://github.com/vector-im/element-web/issues/19304 which are now `:man_with_beard` and `:woman_with_beard:`

commonlawfeature replied:
Ah, got it. I still feel that this method is better, even if not the easiest. The code shows clear intent, and does not rely on assumptions about the content of the strings the function is operating on. It maps very obviously to the code that is adding the surrounding `:`s.

My preference for code with my name on it would be leave it as is, but if it is style for this repo/project or this is the thing stopping this bugfix from getting merged then I will change it immediately.

turt2live replied:
if we're worried about colons in the middle, why not regex instead of manually grabbing the string pieces?

commonlawfeature replied:
Also, I am getting issues with string.replaceAll not existing on the current typescript config. From the tsconfig file, it appears that this project targets es2016. Per https://www.typescriptlang.org/tsconfig#lib, string.replaceAll was added in es2021.

turt2live replied:
I don't think this needs `replaceAll`, just `str.replace(/^:(.*):$/g, "$1")` (untested)

commonlawfeature commented on src/autocomplete/EmojiProvider.tsx:
```diff
@@ -70,7 +88,7 @@ export default class EmojiProvider extends AutocompleteProvider {
         this.matcher = new QueryMatcher<ISortedEmoji>(SORTED_EMOJI, {
             keys: [],
-            funcs: [o => o.emoji.shortcodes.map(s => `:${s}:`)],
+            funcs: [o => o.emoji.shortcodes.map(s => `${EMOJI_DELIMITER}${s}${EMOJI_DELIMITER}`)],
```
Yeah, agreed. I did not have time to look into the other inconsistencies, as they appear to be related to either the search or the scoring. This was just a quick win to get some testing around this area, and fix a particularly annoying but. I am happy to keep looking into the other inconsistencies as well, but I don't think that those fixes, will be in the same area.

MadLittleMods replied:
@commonlawfeature Fixing the main problem is a good iteration 👍 Leave the other stuff for another PR/iteration

turt2live commented on src/autocomplete/EmojiProvider.tsx:
```diff
@@ -62,6 +67,19 @@ function score(query, space) {
     }
 }
 
+function delimiterTrimmed(string: string): string {
```
because our variable names get questionable when they're the same name as the type:
```suggestion
function delimiterTrimmed(str: string): string {
```

commonlawfeature replied:
Done.

turt2live commented on src/autocomplete/EmojiProvider.tsx:
```diff
@@ -33,9 +34,13 @@ import { TimelineRenderingType } from '../contexts/RoomContext';
 
 const LIMIT = 20;
 
+// The delimiter used to start and end emoji shortcodes.
+const EMOJI_DELIMITER = ':';
```
not sure this is safe/best to break out to a constant if we're using it in both regex and strings, as if we were to try and use the value of breaking it out to a variable then we might need to escape the new character in regex but not in strings. 

I'm generally confident that this doesn't need to be a variable though, given the chances of it changing are incredibly slim. It would require a global effort to represent emoji differently for us to change it.

commonlawfeature replied:
Removed.

turt2live reviewed (commented):
overall I think this seems sane - just a few concerns listed in the comments

commonlawfeature commented:
@turt2live I appreciate your review, and believe I have addressed your concerns and responded to your and others feedback. Do you mind helping me understand how else I can help get this bug fixed?

turt2live reviewed (changes requested):
otherwise looking good - just the colon stripping to worry about. Thanks :)

commonlawfeature commented:
Adjusted. Thanks!

turt2live commented:
if you've committed the changes, I'm not seeing them (sorry) - be sure to push :)

commonlawfeature commented:
Apologies. Committed them to an experimental branch by mistake.

turt2live commented:
(please avoid force pushes after a review has been given, as it makes re-review harder)

commonlawfeature commented:
Again, apologies. But, unless I misunderstand something, if I have pushed a commit that is not properly signed-off, my only option is a force push to correct the signoff, right?

turt2live commented:
Signoff can also be provided in the comments or PR description, which saves the signoff in the commit :)

turt2live reviewed (approved):
Thanks!
